Sources/Network/RequestAuthenticator.swift

```diff
@@ -0,0 +1,53 @@
+import Foundation
+import Result
+
+// A type representing a request authenticator.
+public protocol RequestAuthenticator: class {
+
+    /// A type that represents a network request.
+    associatedtype Request
+
+    /// A type that represents a request authentication error.
+    associatedtype Error: Swift.Error
+
+    /// The authenticator's authentication completion closure, invoked when a request's authentication finishes.
+    typealias AuthenticateCompletion = (Result<Request, Error>) -> Cancelable
```

Comment:
Shouldn't we call it `AuthenticaticationCompletion`, as you described in the comment?

Reply:
I guess it can. My train of thought was to name it `AuthenticateCompletion` since the method is `authenticate`, but now that I look at it again it does sound a bit odd 😅

I'm also not completely sold on it being called "completion" (same for `makeRequest`), since it returns a `Cancelable` (meaning itself returns an "operation" that can be cancelled, and should be chained with the one returned from `authenticate`). Any suggestions?